_posts/2015-03-23-design-patterns-part-1.markdown

```diff
@@ -0,0 +1,131 @@
+
+Most Rails apps use the ORM that ships with Rails, ActiveRecord.  To create a user, we'd have something that looks like this.
+
+{% highlight ruby %}
+
+    User.create(first_name: "John", last_name: "Doe", email: "JohnDoe@Example.com")
+
+{% endhighlight %}
+
+At SitterCity, we often want the flexibility to change parts of the app, and that includes changing the ORM.  So, we'll use the repository pattern concentrate all ActiveRecord queries into single class.  Instead of the code you saw above, we might have a class that looks like this:
+
+
+{% highlight ruby %}
+
+class ARUserRepository
+
+  def create(first_name, last_name, email)
+    User.create(first_name: first_name, last_name: last_name, email)
+  end
+
+end
+
+{% endhighlight %}
+
+If we decide that we don't want to use ActiveRecord anymore, there won't be a lot of work involved.  Let's say that we want to use the Sequel ORM.  We'd end up writing another repository that had the exact same interface as the ARUserRepository.  (If you’re unfamiliar with the idea of interfaces, just know that in the world of Ruby, we can say that two classes have the same interface if they have the same public methods available to them.)
+
+{% highlight ruby %}
+
+class SequelUserRepository
+  def create(first_name, last_name, email)
+    $users.insert(first_name: first_name, last_name: last_name, email: email:)
+  end
+end
+
+{% endhighlight %}
+ 
+Or, let's imagine we decide that the data we want to store would fit best in a document store like MongoDB.  We'd just write another repo and as long as it implements the same interface as other repositories, we can swap it in without changing any other parts of the app.
+
+In conjunction with repositories, we also use entities.  An entity is a custom class that possess all the attributes of an object that we care about.  As you know, when you save an ActiveRecord object,  Rails will return you an Active Record object. 
+
+
+{% highlight ruby %}
+
+user = User.new(first_name: "John", last_name: "Doe", email: "JohnDoe@Example.com")
+=> #<User id: nil, first_name: "John", last_name: "Doe", email: "JohnDoe@Example.com")
+user.save
+=> #<User id: 1, first_name: "John", last_name: "Doe", email: "JohnDoe@Example.com")
+
+{% endhighlight %}
+
+Note that we’ve got a new instance of an object that inherits from ActiveRecord::Base and when we save it, we’re still dealing with an object that inherits from ActiveRecord::Base.
+
+But what’s the point of using the generic interface of our repository if we’re going to end up returning AR objects? 
+Entities to the rescue!  We want our app passing around generic entities whose behavior is completely under our control--unlike ActiveRecord instances.  Using ActiveRecord instances couples our code tightly to the Rails framework.  We don't want that.  After we create a resource in our repository, we take the response and translate it into a custom entity.  
+
+Here's an example of what an entity class might look like.
```

Comment:
Thanks for sharing this.  I revised my code to go with something similar to this (though I made it a bit more basic).